### DISCUSSION TITLE
分布式系统 · 分布式事务的实现原理 · /distributed-transaction-principle

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#528] [Announcements]
https://draveness.me/distributed-transaction-principle

**Comment by xueliangliang** [comment]
用的啥画图工具，挺好看的。

**Comment by LonglyCode** [comment]
> @xueliangliang
> 用的啥画图工具，挺好看的。

注意看结尾处: 图片使用 Sketch 进行绘制

**Comment by wangwyx** [comment]
常食 尝试

**Comment by TomyJx** [comment]
这篇文章比较适合对分布式事务有一定了解的再看了，真的连着三篇文章才把分布式事务好好的理解了

**Comment by auxten** [comment]
Hey, 我们在构建一个去中心化区块链特性的数据库 [CovenantSQL](https://github.com/CovenantSQL/CovenantSQL)，很想认识你一下，我邮箱 auxten%covenantsql.io

**Comment by peichanglei109** [comment]
感谢博主分享,另一篇也不错,讲的挺透彻:https://blog.csdn.net/weixin_40533111/article/details/85069536

**Comment by NicePiao** [comment]
hi，作者你好，看了你的一些总结和技术文章，感觉很有帮助。我也是一个从客户端转到服务端的同学，现在在阿里巴巴工作，很希望能够在技术上跟你有一些更多的交流。我的微信号是:  hell0piao0，如果有幸能够加一个微信，将非常荣幸~

**Comment by Imfan** [comment]
“当一个事务常食对数据库进行修改时”
尝试 应该写错了吧

**Comment by draveness** [comment]
> @Imfan
> “当一个事务常食对数据库进行修改时”
> 尝试 应该写错了吧

已经修复了

**Comment by draveness** [comment]
> @NicePiao
> hi，作者你好，看了你的一些总结和技术文章，感觉很有帮助。我也是一个从客户端转到服务端的同学，现在在阿里巴巴工作，很希望能够在技术上跟你有一些更多的交流。我的微信号是:  hell0piao0，如果有幸能够加一个微信，将非常荣幸~
> 

微信上搜不到你

**Comment by gaojunhuicavon** [comment]
作者你好，看了你的文章感觉收获很大，但同时有一点疑问希望能获得你的帮助。想请问saga使用协同方式实现的时候，本地事务执行完毕后，怎么保证一定成功给下游发送一个新的事件呢，如果本地事务完成后，发送新的事件失败该怎么处理？

**Comment by draveness** [comment]
> 作者你好，看了你的文章感觉收获很大，但同时有一点疑问希望能获得你的帮助。想请问saga使用协同方式实现的时候，本地事务执行完毕后，怎么保证一定成功给下游发送一个新的事件呢，如果本地事务完成后，发送新的事件失败该怎么处理？

要先向下游发送事件，成功之后再提交本地事务，下游需要支持重复的消息（接口幂等）

**Comment by gaojunhuicavon** [comment]
> > 作者你好，看了你的文章感觉收获很大，但同时有一点疑问希望能获得你的帮助。想请问saga使用协同方式实现的时候，本地事务执行完毕后，怎么保证一定成功给下游发送一个新的事件呢，如果本地事务完成后，发送新的事件失败该怎么处理？
> 
> 要先向下游发送事件，成功之后再提交本地事务，下游需要支持重复的消息（接口幂等）

如果想下游发送时间后，服务器与数据库发生网络分区，导致无法提交本地事务呢？这个问题也是我一直纠结的问题，希望作者您能给我点启发，十分感谢。

**Comment by draveness** [comment]
> 如果想下游发送时间后，服务器与数据库发生网络分区，导致无法提交本地事务呢？这个问题也是我一直纠结的问题，希望作者您能给我点启发，十分感谢。

那就不提交啊，服务器下次连上数据库的时候，发现有事务状态是 "pending" 就会进行重试，只要下游使用『幂等 ID』保证不会出现问题就行

**Comment by Sunnykaby** [comment]
3pc有点事情没有太明白，虽然增加了一个中间步骤，但是can阶段已经把资源lock住了是么？但是这不是变相的又增加了风险么，增加一次网络调用：1，耗时增加；2，出问题的概率增大？

**Comment by draveness** [comment]
> 3pc有点事情没有太明白，虽然增加了一个中间步骤，但是can阶段已经把资源lock住了是么？但是这不是变相的又增加了风险么，增加一次网络调用：1，耗时增加；2，出问题的概率增大？

它们在处理错误时的行为不太一样，2pc 可能会有部分参与者完成提交、部分无法提交，但是正确实现的 3pc 通过 preCommit 可以预先确定参与者是否能提交，不会出现部分提交的不一致问题。

> The three-phase commit protocol eliminates this problem by introducing the Prepared to commit state. If the coordinator fails before sending preCommit messages, the cohort will unanimously agree that the operation was aborted. The coordinator will not send out a doCommit message until all cohort members have ACKed that they are Prepared to commit. This eliminates the possibility that any cohort member actually completed the transaction before all cohort members were aware of the decision to do so (an ambiguity that necessitated indefinite blocking in the two-phase commit protocol).
>
> <https://en.wikipedia.org/wiki/Three-phase_commit_protocol>
